### DISCUSSION TITLE
[JS101] 關於安安++和if else的疑惑

### DISCUSSION THREAD

**Discussion by ming320003** [Lidemy/forum#38] [前端相關問題]
標題：關於安安++和if/else的疑惑
內文：
在看影片時，有覺得奇怪的地方，也不知道是不是我自己理解錯了？

1. 想請問在「安安，可以 ++ 嗎」單元的 3:15 是不是有口誤？(因為影片這個段落，如果單純聽語音的話，還是在按照過去的邏輯運算的方法來看的階段，是到後面才開始說++放前放後的順序問題，但是老師好像在結果上就已經把++前後考慮進來了)

2. 在「如果我說，愛我沒有如果：if/else statement」單元的 9:40 ，為甚麼 1 會是 5 的倍數？
謝謝

**Comment by aszx87410** [comment]
> 想請問在「安安，可以 ++ 嗎」單元的 3:15 是不是有口誤？(因為影片這個段落，如果單純聽語音的話，還是在按照過去的邏輯運算的方法來看的階段，是到後面才開始說++放前放後的順序問題，但是老師好像在結果上就已經把++前後考慮進來了)

我看了一下，這邊想強調的應該是之前在講 ++ 的時候我把他解釋為跟「a+=1」是等價的，如果這個是真的的話，那 `console.log(a++ && 30)` 應該要是 30。但執行後卻發現結果是 0，代表在執行那一行的時候 a 是 0，所以 a++ 跟 a+=1 是不等價的，把他們跟其他敘述放在同一行的時候會有執行優先順序的問題，而這這個優先順序就是 a++ 跟 ++a 的不同點

> 在「如果我說，愛我沒有如果：if/else statement」單元的 9:40 ，為甚麼 1 會是 5 的倍數？

這一段想強調的是，如果 number % 5 === 0 的話，那是不是代表 `!(number % 5)` 就會是 1？
其實這就是把 `number % 5 === 0` 左右兩邊都加上 ! 的感覺，就變成 `!(number % 5) === 1` 了

所以是這整句是 1，而不是代表 1 是 5 的倍數

**Reply by ming320003** [reply]
謝謝老師的解答。我後來又回去看了安安++的影片，然後發現了新的疑惑。
老師在影片最後面說程式還不如寫成
console.log (a && 30)
a++;
然後老師輸出的結果是0。

但是依照我的想法，a && 30 答案會是0，之後的a++不是應該要讓輸出的答案變成1嗎？
我覺得很奇怪，於是想說是不是因為a++沒有用console.log導致它輸不出來，於是幫它補上了console.log，結果輸出的卻是0和0，而不是我以為的0和1。
var a =0
console.log (a && 30)
console.log ( a++)



後來我又試了另一個寫法，依然不行。
var a = 0
a && 30
console.log ( a++)

想請教老師我究竟是哪裡出錯了呢？謝謝老師

**Reply by aszx87410** [reply]
console.log 輸出的是當下的值，然後沒有 console.log 就不會輸出東西，所以「於是想說是不是因為a++沒有用console.log導致它輸不出來」這是正確的

然後看起來你把問題弄複雜了，你先看這個範例就好：

``` js
var a = 0
console.log(a++)
console.log(a)
```

這會輸出什麼？先輸出 0，然後是 1
因為之前有提過了 ++ 放後面的話會在整行的東西跑完以後才執行 ++，所以執行到 `console.log(a++)` 的時候，會先輸出 a，然後再把 a+1
所以會輸出 0
然後第二個 console.log 的時候會輸出 1，因為那時候 a 已經變成 1 了

> 但是依照我的想法，a && 30 答案會是0，之後的a++不是應該要讓輸出的答案變成1嗎？

前面提過了，console.log 輸出的是當下的值，所以如果你是這樣寫：

``` js
var a = 0;
console.log (a && 30) // 0
a++;
console.log(a) // 1
```

就會是你期望的狀況了

有幾件事情你可以再想一下：

1. console.log 是做什麼的
2. ++ 放前面跟後面的差別

**Reply by ming320003** [reply]
謝謝老師的回答，我覺得我懂了！！

**Comment by NewmanJian** [comment]
老師
我看到相似的範例如下
let a = 1, b = 1;
b = a++ ;
console.log(a, b) // 2,1

因為++在後面，所以會先把值賦予後再加1，b <= a++，b=a，印出b=1 ，得證。
但我不懂的是為什麼印出a=2，而不是a=1，a++不是已經賦予給b了嗎？
我哪個步驟出錯了？

**Reply by allenliaooooooo** [reply]
`b = a++;`
是把 a 的值賦給 b 後，`a += 1`

**Reply by NewmanJian** [reply]
阿，沒有注意到a賦值給b後，還要繼續跑 a=a+1，這樣子就解釋的通了，謝謝同學 : )
